src/NetTopologySuite.IO.Esri/Readers/ShapefileReader.cs

```diff
@@ -0,0 +1,296 @@
+        /// </summary>
+        public Envelope BoundingBox { get; } = new Envelope();
+
+
+        /// <summary>
+        /// Reads geometry and feature attributes from underlying SHP and DBF files. 
+        /// </summary>
+        /// <param name="geometry">Feature geometry.</param>
+        /// <param name="attributes">Feature atrributes.</param>
+        /// <param name="deleted">Indicates if the record was marked as deleted.</param>
+        /// <returns>
+        /// true if the enumerator was successfully advanced to the next record;
+        /// false if the enumerator has passed the end of the table.
+        /// </returns>
+        public abstract bool Read(out Geometry geometry, out AttributesTable attributes, out bool deleted);
+
+
+        /// <inheritdoc/>
+        protected override void DisposeManagedResources()
+        {
+            Reader?.Dispose();
+            base.DisposeManagedResources(); // This will dispose streams used by ShpReader and DbfReader. Do it at the end.
+        }
+
+        internal void Restart()
+        {
+            Reader.Restart();
+        }
+
+        #region *** Enumerator ***
+
+        IEnumerator<Feature> IEnumerable<Feature>.GetEnumerator()
+        {
+            return new FeatureEnumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new FeatureEnumerator(this);
+        }
+
+        private class FeatureEnumerator : IEnumerator<Feature>
+        {
+            private readonly ShapefileReader Owner;
+            public Feature Current { get; private set; }
+            object IEnumerator.Current => Current;
+
+            public FeatureEnumerator(ShapefileReader owner)
+            {
+                Owner = owner;
+            }
+
+            public void Reset()
+            {
+                Owner.Restart();
+            }
+
+            public bool MoveNext()
+            {
+                if (!Owner.Read(out var geometry, out var attributes, out var deleted))
+                {
+                    return false;
+                }
+
+                if (deleted)
+                {
+                    return MoveNext();
+                }
+
+                Current = geometry.ToFeature(attributes); 
+                return true;
+            }
+
+            public void Dispose()
+            {
+                // Nothing to dispose
+            }
+        }
+
+        #endregion
+
+
+        /// <summary>
+        /// Opens shapefile reader.
+        /// </summary>
+        /// <param name="shpPath">Path to shapefile.</param>
+        /// <param name="encoding">DBF file encoding. If null encoding will be guess from related .CPG file or from reserved DBF bytes.</param>
+        /// <returns>Shapefile reader.</returns>
+        public static ShapefileReader Open(string shpPath, Encoding encoding = null)
+        {
+            shpPath = Path.ChangeExtension(shpPath, ".shp");
```

Comment:
Some logic to query SRID from accompanied *.prj file would be nice. Use it to set up `GeometryFactory` for ShapefileReaders. If not just set `0`.
```
int srid = GetSRID(System.IO.Path.ChangeExtension(shpPath, "prj");
var factory = NtsGeometryServices.Instance.CreateGeometryFactory(srid);
```